src/accumulator/proof.rs

```diff
@@ -241,6 +243,270 @@ impl Proof {
+                new_proofs.push((new_pos, *hash));
+            }
+            return new_proofs;
+        }
+
+        positions
+    }
+
+    /// update_proof_remove modifies the cached proof with the deletions that happen in the block proof.
+    /// It updates the necessary proof hashes and un-caches the targets that are being deleted.
+    fn update_proof_remove(
+        self,
+        block_targets: Vec<u64>,
+        cached_hashes: Vec<sha256::Hash>,
+        updated: Vec<(u64, sha256::Hash)>,
+        num_leaves: u64,
+    ) -> Result<(Proof, Vec<sha256::Hash>), String> {
+        let total_rows = util::tree_rows(num_leaves);
+
+        let targets_with_hash: Vec<(u64, bitcoin_hashes::sha256::Hash)> = self
+            .targets
+            .clone()
+            .into_iter()
+            .zip(cached_hashes.clone().into_iter())
+            .filter(|(pos, _)| !block_targets.contains(pos))
+            .collect();
+        let (targets, target_hashes) = targets_with_hash.to_owned().into_iter().unzip();
+
+        let proof_positions =
+            util::get_proof_positions(&self.targets, num_leaves, util::tree_rows(num_leaves));
+
+        let old_proof: Vec<_> = proof_positions
+            .iter()
+            .zip(self.hashes.clone().into_iter())
+            .collect();
+
+        let mut new_proof = vec![];
+        // Grab all the positions of the needed proof hashes.
+        let needed_pos = util::get_proof_positions(&targets, num_leaves, total_rows);
+
+        let updated: Vec<_> = updated
+            .into_iter()
+            .filter(|(pos, _)| proof_positions.contains(pos))
+            .collect();
+
+        let old_proof_iter = old_proof.into_iter();
+
+        // Loop through old_proofs and only add the needed proof hashes.
+        for (pos, hash) in old_proof_iter {
+            if let Some((_, updated_hash)) =
+                updated.iter().find(|(updated_pos, _)| pos == updated_pos)
+            {
+                if *updated_hash != sha256::Hash::default() {
+                    new_proof.push((*pos, *updated_hash));
+                }
+            } else {
+                if hash != sha256::Hash::default() {
+                    new_proof.push((*pos, hash));
+                }
+            }
+        }
+
+        let missing_positions = needed_pos
+            .into_iter()
+            .filter(|pos| proof_positions.contains(pos))
+            .peekable();
+
+        for missing in missing_positions {
+            if let Some((_, hash)) = updated
+                .iter()
+                .find(|(updated_pos, _)| missing == *updated_pos)
+            {
+                if *hash != sha256::Hash::default() {
+                    new_proof.push((missing, *hash));
+                }
+            }
+        }
+        let (_, hashes): (Vec<u64>, Vec<sha256::Hash>) = new_proof.into_iter().unzip();
+        let targets =
+            Proof::calc_next_positions(&block_targets, &targets_with_hash, num_leaves, false)?;
```

Comment:
`calc_next_positions()` should also be called on the proofs since they'll also change positions.

Here's a test case that makes the current code fail.

```
    {
        "update": {
            "adds": [],
            "proof": {
                "targets": [4, 5, 6, 7],
                "hashes": [
                    "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386"
                ]
            },
            "del_hashes": [
                "e52d9c508c502347344d8c07ad91cbd6068afc75ff6292f062a09ca381c89e71",
                "e77b9a9ae9e30b0dbdb6f510a264ef9de781501d7b6b92ae89eb059c5ab743db",
                "67586e98fad27da0b9968bc039a1ef34c939b9b8e523a8bef89d478608c5ecf6",
                "ca358758f6d27e6cf45272937977a748fd88391db679ceda7dc7bf1f005ee879"
            ]
        },
        "remembers": [],
        "cached_proof": {
            "targets": [1, 8, 12],
            "hashes": [
                "6e340b9cffb37a989ca544e6bb780a2c78901d3fb33738768511a30617afa01d",
                "2b4c342f5433ebe591a1da77e013d1b72475562d48578dca8b84bac6651c3cb9",
                "9d1e0e2d9459d06523ad13e28a4093c2316baafe7aec5b25f30eba2e113599c4",
                "9576f4ade6e9bc3a6458b506ce3e4e890df29cb14cb5d3d887672aef55647a2b",
                "c413035120e8c9b0ca3e40c93d06fe60a0d056866138300bb1f1dd172b4923c3",
                "29590a14c1b09384b94a2c0e94bf821ca75b62eacebc47893397ca88e3bbcbd7"
            ]
        },
        "initial_stump": "Add 14 leaves [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13]",
        "initial_roots": [
            "b151a956139bb821d4effa34ea95c17560e0135d1e4661fc23cedc3af49dac42",
            "9c053db406c1a077112189469a3aca0573d3481bef09fa3d2eda3304d7d44be8",
            "55d0a0ef8f5c25a9da266b36c0c5f4b31008ece82df2512c8966bddcc27a66a0"
        ],
        "initial_leaves": 14,
        "cached_hashes": [
            "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a",
            "beead77994cf573341ec17b58bbf7eb34d2711c993c1d976b128b3188dc1829a",
            "ef6cbd2161eaea7943ce8693b9824d23d1793ffb1c0fca05b600d3899b44c977"
        ],
        "expected_roots": [
            "df46b17be5f66f0750a4b3efa26d4679db170a72d41eb56c3e4ff75a58c65386",
            "9c053db406c1a077112189469a3aca0573d3481bef09fa3d2eda3304d7d44be8",
            "55d0a0ef8f5c25a9da266b36c0c5f4b31008ece82df2512c8966bddcc27a66a0"
        ],
        "expected_targets": [8, 12, 17],
        "expected_cached_hashes": [
            "beead77994cf573341ec17b58bbf7eb34d2711c993c1d976b128b3188dc1829a",
            "ef6cbd2161eaea7943ce8693b9824d23d1793ffb1c0fca05b600d3899b44c977",
            "4bf5122f344554c53bde2ebb8cd2b7e3d1600ad631c385a5d7cce23c7785459a"
        ]
    }
```

Reply:
Right, if a proof node gets moved up, it might make the proof in the wrong order. Pushing a fix.